src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
```

Comment:
Name should be: ConvertQueryToForEach

Reply:
I thought about using this class as a base class for QueryToForEach, ForEachToQuery, QueryToLINQMethod and so on, for LINQ related conversion. That is the reason for the name.
Will add comment about this.

---
In reply to: [175267977](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267977) [](ancestors = 175267977)

Reply:
I would be very surprised if all those classes could have any common logic.  They will all operate so differently.

Reply:
yes, I think that code currently provided in AbstractConvertLinqProvider can be shared between various LINQ Query, LINQ method, foreach conversions.

---
In reply to: [175580214](https://github.com/dotnet/roslyn/pull/25362#discussion_r175580214) [](ancestors = 175580214)

Reply:
Sorry, i'm just not seeing it.  All the code that exists seems to just be:

1. a check for the misc workspace (not sure why we need this).
2.  getting the semanticmodel.
3. getting syntaxfacts (which are unused).

There's also this analyzer type that gets created.  But i don't actually see why that guy is necessary.  The CodeRefactoringProvider is already the thing that does the analysis.

The only common stuff i see here is a pattern to convert CodeRefactoringProvider from being the one that does work, to having a system where there is an Analyzer, and it still does nothing, requiring the subclasses to do all that work.

>  foreach conversions.

So, for example, i just did the for-to-foreach conversion.  But i didn't find any need to follow this analyzer approach.  It seems odd to create this feeling that that's the approach that would then be required for all these subclasses.

Reply:
I still need an internal class to propogate references to semanticModel and cancellationToken across the algorithm. I can move it into a private class within CSharpConvertLinqQueryToLinqMethodProvider

---
In reply to: [176513680](https://github.com/dotnet/roslyn/pull/25362#discussion_r176513680) [](ancestors = 176513680,176267105)

Reply:
Why can't those be passed as method parameters?

---
In reply to: [176528574](https://github.com/dotnet/roslyn/pull/25362#discussion_r176528574) [](ancestors = 176528574,176513680,176267105)

Reply:
Note: i'm fine with the inner class approach.  I'm sympathetic to not wanting to pass the same set of params to all methods (especially for a feature as large as this).  That said, i would have each individual feature decide which data they want to package up and have them all make their own nested class.  Having all the features try to agree that they must have this IAnalyzer type and whatnot seems of little benefit.

Reply:
agree. thank you!

---
In reply to: [176577464](https://github.com/dotnet/roslyn/pull/25362#discussion_r176577464) [](ancestors = 176577464)